beetmoverscript/src/beetmoverscript/script.py

```diff
@@ -285,6 +290,206 @@ async def upload_data(context):
+               "dest1",
+             ]
+           }
+         },
+       },
+       {
+         "taskId": "taskId2",
+         "paths": {
+           "*.log": {
+             "destinations": [
+               "logdest",
+             ]
+           },
+           "*": {
+             "destinations": [
+               "dest1",
+             ]
+           }
+         },
+       }
+     ]
+
+     The return value takes the form of:
+     [
+         {
+             "taskId": "taskId1",
+             "paths": {
+                 "/path/to/public/build/file": {
+                     "destinations": [
+                         "dest1/file",
+                     ]
+                 },
+                 "/path/to/public/build/file2": {
+                     "destinations": [
+                         "dest1/file2",
+                     ]
+                 },
+                 "/path/to/public/logs/live.log": {
+                     "destinations": [
+                         "dest1/live.log",
+                     ]
+                 },
+             }
+         },
+         {
+             "taskId": "taskId2",
+             "paths": {
+                 "/path/to/public/build/file3": {
+                     "destinations": [
+                         "dest1/file3",
+                     ]
+                 },
+                 "/path/to/public/logs/live.log": {
+                     "destinations": [
+                         "logdest/live.log",
+                     ]
+                 },
+             }
+         }
+     ]
+
+     Each artifact may match up to two `paths` in the `artifactMap`: a `*` path
+     and one other path. If an artifact matches both of these, the non-`*` one
+     takes precedence, and it will _not_ be written to the `*` path destinations.
+     This allows us to have specific filenames (or extensions, prefixes, etc.)
+     grouped together, and to have a "catch-all" for any unmatched artifacts.
+
+     Sanity checking is done to ensure that no destination files would overwrite
+     each other.
+    """
+
+    concreteArtifactMap = []
+    errors = []
+
+    for map_ in artifactMap:
+        concretePaths = {}
+
+        full_glob_destinations = map_["paths"].get("*", {}).get("destinations")
+        other_paths = copy.deepcopy(map_["paths"])
+        if "*" in other_paths:
+            del other_paths["*"]
+
+        for taskId, artifacts in upstreamArtifactPaths.items():
+            if map_["taskId"] != taskId:
+                continue
+
+            for artifact in artifacts:
+                # the artifact path with the scriptworker-specific leading
+                # directories removed. ie: in the same form it is in the task
+                # it was pulled from
+                leading_dir = os.path.join(work_dir, "cot", taskId) + "/"
+                if not artifact.startswith(leading_dir):
+                    raise ScriptWorkerTaskException("cannot determine relative artifact path")
```

Comment:
suggestion: include `artifact` in the exception message so it's easier to figure out if this ever happens?